### PR TITLE
Support for test/variant data files

### PR DESCRIPTION
**Pull request by clebergnu**
This introduces support for test/variant data files, by deprecating the `datadir` property, and introducing `get_data()` API.

Also, a number of misc fixes and improvements collected along the way.

### CODE REVIEW

**Comment by clebergnu** [comment]
[Test's .data directory should include the Variant-ID](https://trello.com/c/wiGkOFSa/828-tests-data-directory-should-include-the-variant-id)

File: selftests/unit/test_xunit.py [hunk 1]
```diff
@@ -56,7 +56,7 @@ def tearDown(self):
         os.remove(self.tmpfile[1])
         shutil.rmtree(self.tmpdir)
 
-    def testAddSuccess(self):
+    def test_add_success(self):
```
Comment (apahim) [inline selftests/unit/test_xunit.py]: Ok, only occurrences I see left are in the test script codes present in `unit/test_loader.py`, but I don't think we should touch that.

File: avocado/utils/astring.py [hunk 2]
```diff
@@ -209,25 +209,30 @@ def tabular_output(matrix, header=None):
...
+
+    :param input_str: String to be converted
     :return: String which is safe to pass as a file/dir name (on recent fs)
     """
-    if string.startswith("."):
-        string = "_" + string[1:]
-    return string.replace(os.path.sep, '_')[:255]
+    if input_str.startswith("."):
+        input_str = "_" + input_str[1:]
+    return input_str.replace(os.path.sep, '_')[:255]
```
Comment (apahim) [inline avocado/utils/astring.py]: Ok, change is welcome and harmless.

File: avocado/core/test.py [hunk 3]
```diff
@@ -929,7 +929,7 @@ def test(self):
                                new_cwd)
                 os.chdir(new_cwd)
 
-            self.execute_cmd()
+            self._execute_cmd()
```
Comment (apahim) [inline avocado/core/test.py]: Ok, good one.

File: docs/source/WritingTests.rst [hunk 4]
```diff
@@ -338,14 +338,14 @@ values). In total it'll produce 8 variants of each test::
...
-       (7/8) passtest.py:PassTest.test;7: PASS (0.01 s)
-       (8/8) passtest.py:PassTest.test;8: PASS (0.01 s)
+       (1/8) passtest.py:PassTest.test;builtin-one_cycle-f659: PASS (0.01 s)
+       (2/8) passtest.py:PassTest.test;builtin-six_cycles-723b: PASS (0.01 s)
+       (3/8) passtest.py:PassTest.test;builtin-one_hundred_cycles-633a: PASS (0.01 s)
+       (4/8) passtest.py:PassTest.test;builtin-six_hundred_cycles-a570: PASS (0.01 s)
+       (5/8) passtest.py:PassTest.test;shell-one_cycle-55f5: PASS (0.01 s)
+       (6/8) passtest.py:PassTest.test;shell-six_cycles-9e23: PASS (0.01 s)
+       (7/8) passtest.py:PassTest.test;shell-one_hundred_cycles-586f: PASS (0.01 s)
+       (8/8) passtest.py:PassTest.test;shell-six_hundred_cycles-1e84: PASS (0.01 s)
```
Comment (apahim) [inline docs/source/WritingTests.rst]: yep

File: docs/source/ReferenceGuide.rst [hunk 5]
```diff
@@ -111,7 +111,6 @@ Syntax::
 Examples of test-names::
 
    '/bin/true'
-   '/bin/grep foobar /etc/passwd'
```
Comment (apahim) [inline docs/source/ReferenceGuide.rst]: Fine-tooth comb mode On :)

File: docs/source/ReferenceGuide.rst [hunk 6]
```diff
@@ -108,14 +116,13 @@ Syntax::
...
-   'file:///tmp/passtest.py:Passtest.test'
-   'multiple_tests.py:MultipleTests.test_hello'
-   'type_specific.io-github-autotest-qemu.systemtap_tracing.qemu.qemu_free'
+Example of Test IDs::
 
+   '001-/bin/true'
+   '002-passtest.py:Passtest.test;quiet-'
+   '003-file:///tmp/passtest.py:Passtest.test'
+   '004-multiple_tests.py:MultipleTests.test_hello;maximum_debug-df2f'
+   '005-type_specific.io-github-autotest-qemu.systemtap_tracing.qemu.qemu_free'
```
Comment (apahim) [inline docs/source/ReferenceGuide.rst]: ok, only point here is that `001` will only happen if the job has between 100 and 999 tests. For this example, I'd keep it coherent with the number of tests you're demonstrating...
```
'1-/bin/true'
'2-passtest.py:Passtest.test;quiet-'
...
```

File: avocado/core/data_dir.py [hunk 7]
```diff
@@ -109,7 +104,8 @@ def get_test_dir():
         return configured
 
     if settings.settings.intree:
-        return _IN_TREE_TESTS_DIR
+        base_dir = os.path.dirname(os.path.dirname(os.path.dirname(__file__)))
+        return os.path.join(base_dir, 'examples', 'tests')
```
Comment (apahim) [inline avocado/core/data_dir.py]: ok, good one.

File: avocado/core/test.py [hunk 8]
```diff
@@ -73,12 +73,12 @@ def __init__(self, uid, name, variant=None, no_digits=None):
...
-        self.variant = variant["variant_id"] if variant else None
-        if variant is None:
+        if variant is None or variant["variant_id"] is None:
+            self.variant = None
             self.str_variant = ""
         else:
-            self.str_variant = (";" + str(variant["variant_id"])
-                                if variant["variant_id"] else "")
+            self.variant = variant["variant_id"]
+            self.str_variant = ";%s" % self.variant
```
Comment (apahim) [inline avocado/core/test.py]: Nice optimization. +1.

File: avocado/core/test.py [hunk 9]
```diff
@@ -52,17 +52,6 @@
...
-
-    """
-    The given test name is not a TestName instance
-
-    With the introduction of :class:`avocado.core.test.TestName`, it's
-    not allowed to use other types as the ``name`` parameter to a test
-    instance.  This exception is raised when this is attempted.
-    """
-
-
```
Comment (apahim) [inline avocado/core/test.py]: Up to here I can not tell why this is needed, but my guess is that I will discover soon, in the next commits. Maybe this change should come together with the code that actually uses it? Or, if that's not the case, a better context in the commit message would suffice.

File: docs/source/ReferenceGuide.rst (hunk 6 above)
Comment (clebergnu) [inline docs/source/ReferenceGuide.rst]: True... I'll change that.

File: avocado/core/test.py (hunk 9 above)
Comment (clebergnu) [inline avocado/core/test.py]: Oh, the point here is that in Python we should not be doing type check/enforcement.  This bit was here temporarily, when the `name` parameter to `avocado.core.test.Test` got changed.  I'll add a better commit message.

File: avocado/core/test.py [hunk 10]
```diff
@@ -248,7 +265,7 @@ def record_and_warn(*args, **kwargs):
     @property
     def name(self):
         """
-        The test name (TestName instance)
+        FIXME: The test name
```
Comment (clebergnu) [inline avocado/core/test.py]: Oops, my bad here.

File: avocado/core/test.py [hunk 11]
```diff
@@ -72,7 +72,7 @@ def __init__(self, uid, name, variant=None, no_digits=None):
             self.str_uid = str(uid).zfill(no_digits if no_digits else 3)
         else:
             self.str_uid = str(uid)
-        self.name = name or "<unknown>"
+        self.name = name
```
Comment (apahim) [inline avocado/core/test.py]: Before the change:
```
>>> test.TestName(0, None)
>>> '0-<unknown>'
```
After the change:
```
>>> test.TestName(0, None)
>>> '0-None'
```
You're basically exchanging a human-friendly string for a machine readable value converted to string, but nothing really changes from the abuse-protection perspective. How exactly do you plan to (hopefully) flag abuses? :)

Comment (apahim) [inline avocado/core/test.py]: Before the change:
```
>>> test.TestName(0, None)
>>> '0-<unknown>'
```
After the change:
```
>>> test.TestName(0, None)
>>> '0-None'
```
You're basically exchanging a human-friendly string for a machine readable value converted to string, but nothing really changes from the abuse-protection perspective. How exactly do you plan to (hopefully) flag abuses? :)

Comment (clebergnu) [inline avocado/core/test.py]: Good point.  I think I had the intentions to do something extra here (or not).  Let me think about it again.

File: avocado/core/test.py [hunk 12]
```diff
@@ -92,26 +96,39 @@ def __eq__(self, other):
...
+        have to shorten either the Test Name or the Variant ID.
+
+        The first component of a Test ID, the numeric unique test id,
+        AKA "uid", will be used as a an stable identifier between the
+        Test ID and the file or directory created based on the return
+        value of this method.  If the filesystem can not even
+        represent the "uid", than an exception will be raised.
+
+        For Test ID "001-mytest;foo", examples of shortened file
+        system versions include "001-mytest;f" or "001-myte;foo".
```
Comment (clebergnu) [inline avocado/core/test.py]: I noticed that this is actually incorrect.  When a Test ID contains a variant, the shortening will happen initially on the test name *only*. Only after it is exhausted (test name is removed), then the variant id will be shortened.

File: avocado/core/test.py [hunk 13]
```diff
@@ -52,18 +52,22 @@
...
     def __init__(self, uid, name, variant=None, no_digits=None):
         """
-        Test name according to avocado specification
+        Constructs a TestID instance
 
         :param uid: unique test id (within the job)
-        :param name: test name (identifies the executed test)
+        :param name: test name, as returned by the Avocado test resolver
+                     (AKA as test loader)
         :param variant: variant id
```
Comment (apahim) [inline avocado/core/test.py]: It's not the `variant id` that will be received in this param, but the variant itself. Not this commit fault, but might be this commit fix :)

File: selftests/unit/test_test.py [hunk 14]
```diff
@@ -245,5 +245,81 @@ def tearDown(self):
...
+        name = 'test'
+        variant_id = 'fast_' * 51    # 255 characters
+        variant = {'variant_id': variant_id}
+        test_id = test.TestID(uid, name, variant=variant)
+        self.assertEqual(test_id.uid, 1)
+        self.assertEqual(test_id.str_filesystem, '%s;%s' % (uid, variant_id[:253]))
+        self.assertIs(test_id.variant, variant_id)
+        self.assertEqual(test_id.str_variant, ";%s" % variant_id)
+
+
```
Comment (apahim) [inline selftests/unit/test_test.py]: ok, cool.

File: avocado/core/test.py (hunk 13 above)
Comment (clebergnu) [inline avocado/core/test.py]: True, I'll fix that.

File: avocado/core/test.py [hunk 15]
```diff
@@ -341,6 +341,54 @@ def datadir(self):
...
+                    return variant_filename
+
+        test_filename = os.path.join(test_datadir, filename)
+        if os.path.exists(test_filename):
+            return test_filename
+
+        file_filename = os.path.join(datadir, filename)
+        if os.path.exists(file_filename):
+            return file_filename
+        return None
```
Comment (apahim) [inline avocado/core/test.py]: Ok, here's where the magic happens. Question:
I am an user who wants to put different files in the data directories of my variants. How do I know the path I should create to put my file on? Today, printing `self.datadir` I see the path avocado will look for. With the new `get_data()`, I have no idea. Will we rely on documentation only? Maybe a `self.list_data()` might help? Not sure...

Comment (clebergnu) [inline avocado/core/test.py]: The first step is good documentation.  If Avocado is clear enough about what a test name is, a variant name is, etc, users *may* be able to figure out on their own.

Then, on top of that, I though of adding debug messages similar to what `self.params.get()` do, showing that while looking for a given data file, the following paths were searched (and would be searched).

Finally, we could have a tool/option that would list and/or create the skeleton data dir for the tests.

```
$ avocado run --create-test-data-dirs test.py
```

Would suit me well.  Thoughts and ideas?

Comment (apahim) [inline avocado/core/test.py]: The problem with the docs-only approach is that the variant id is unpredictable, so user need at least `avocado run` (or `avocado variants`) to figure it out.
With that said, +1 for the log and +1 for the option, but I'd instead add an option to the run command, like `avocado run mytest.py -m myvariants.yaml --create-test-data-dirs --dry-run`.

Comment (clebergnu) [inline avocado/core/test.py]: Yep, I actually edited my previous response, and the first version included a `run` command line option instead.  So yes, I think that's a good option.

File: selftests/unit/test_xunit.py (hunk 1 above)
Comment (clebergnu) [inline selftests/unit/test_xunit.py]: Yep, that was intentional.

**Pull request closed without merging**